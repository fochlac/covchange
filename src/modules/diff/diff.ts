export function diffReports(original: Core.Report, changed: Core.Report): Core.DiffReport {
	const deletedFiles = Object.keys(original.files).filter(name => !changed.files[name])
	const changedFiles = Object.keys(original.files).filter(name => {
		const altered = changed.files[name] && changed.files[name].metrics
		const orig = original.files[name].metrics
		if (!altered) {
			return false
		}

		return altered && (altered.statements !== orig.statements || altered.coveredStatements !== orig.coveredStatements)
	})
	const newFiles = Object.keys(changed.files).filter(name => !original.files[name])

	return {
		total: diffMetrics(original.metrics, changed.metrics),
		deleted: deletedFiles,
		changed: changedFiles.reduce((files, name) => {
			files[name] = diffMetrics(original.files[name].metrics, changed.files[name].metrics)
			return files
		}, {}),
		new: newFiles.reduce((files, name) => {
			files[name] = changed.files[name].metrics
			return files
		}, {}),
	}
}

function diffMetrics(original: Core.Metrics, changed: Core.Metrics): Core.Diff {
	return {
		original,
		changed,
		diff: {
			statements: Math.floor((changed.statements - original.statements) * 10) / 10,
			conditionals: Math.floor((changed.conditionals - original.conditionals) * 10) / 10,
			coveredStatements: Math.floor((changed.coveredStatements - original.coveredStatements) * 10) / 10,
			coveredConditionals: Math.floor((changed.coveredConditionals - original.coveredConditionals) * 10) / 10,
			statementCov: Math.floor((changed.statementCov - original.statementCov) * 10) / 10,
			conditionalCov: Math.floor((changed.conditionalCov - original.conditionalCov) * 10) / 10,
		},
	}
}
