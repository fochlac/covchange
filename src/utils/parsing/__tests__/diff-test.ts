import { mockReport, mockReport2 } from '../../../__tests__/data/reports'

import { diffReports } from '../diff'

describe('diff-module', () => {
	it('should find the diff between the reports', () => {
		expect(diffReports(mockReport, mockReport2)).toMatchSnapshot()
	})
})
