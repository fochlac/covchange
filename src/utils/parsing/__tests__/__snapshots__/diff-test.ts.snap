// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff-module should find the diff between the reports 1`] = `
Object {
  "changed": Object {
    "orders/app/components/order-tours-tab/order-tours-table-renderer.js": Object {
      "changed": Object {
        "conditionalCov": 94.29,
        "conditionals": 35,
        "coveredConditionals": 33,
        "coveredStatements": 44,
        "statementCov": 97.78,
        "statements": 45,
      },
      "diff": Object {
        "conditionalCov": 94.2,
        "conditionals": 0,
        "coveredConditionals": 33,
        "coveredStatements": 44,
        "statementCov": 97.7,
        "statements": 10,
      },
      "original": Object {
        "conditionalCov": 0,
        "conditionals": 35,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 35,
      },
    },
    "orders/app/components/order-tours-tab/order-tours-table.js": Object {
      "changed": Object {
        "conditionalCov": 85.37,
        "conditionals": 41,
        "coveredConditionals": 35,
        "coveredStatements": 22,
        "statementCov": 81.48,
        "statements": 27,
      },
      "diff": Object {
        "conditionalCov": 85.3,
        "conditionals": 0,
        "coveredConditionals": 35,
        "coveredStatements": 22,
        "statementCov": 81.4,
        "statements": 0,
      },
      "original": Object {
        "conditionalCov": 0,
        "conditionals": 41,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 27,
      },
    },
    "orders/app/components/orders-filter-bar.js": Object {
      "changed": Object {
        "conditionalCov": 50,
        "conditionals": 22,
        "coveredConditionals": 11,
        "coveredStatements": 53,
        "statementCov": 96.36,
        "statements": 55,
      },
      "diff": Object {
        "conditionalCov": 18.1,
        "conditionals": 0,
        "coveredConditionals": 4,
        "coveredStatements": -3,
        "statementCov": 12.7,
        "statements": -12,
      },
      "original": Object {
        "conditionalCov": 31.82,
        "conditionals": 22,
        "coveredConditionals": 7,
        "coveredStatements": 56,
        "statementCov": 83.58,
        "statements": 67,
      },
    },
    "orders/app/components/orders-view.js": Object {
      "changed": Object {
        "conditionalCov": 100,
        "conditionals": 17,
        "coveredConditionals": 17,
        "coveredStatements": 44,
        "statementCov": 93.62,
        "statements": 47,
      },
      "diff": Object {
        "conditionalCov": 0,
        "conditionals": 0,
        "coveredConditionals": 0,
        "coveredStatements": -3,
        "statementCov": -6.4,
        "statements": 0,
      },
      "original": Object {
        "conditionalCov": 100,
        "conditionals": 17,
        "coveredConditionals": 17,
        "coveredStatements": 47,
        "statementCov": 100,
        "statements": 47,
      },
    },
    "orders/app/index.js": Object {
      "changed": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "coveredConditionals": 0,
        "coveredStatements": 1,
        "statementCov": 7.69,
        "statements": 13,
      },
      "diff": Object {
        "conditionalCov": 0,
        "conditionals": 0,
        "coveredConditionals": 0,
        "coveredStatements": 1,
        "statementCov": 7.6,
        "statements": 1,
      },
      "original": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 12,
      },
    },
  },
  "deleted": Array [
    "orders/module.js",
  ],
  "new": Object {
    "orders/app/components/order-tours-tab/order-tours-table-core.js": Object {
      "conditionalCov": 71.43,
      "conditionals": 7,
      "coveredConditionals": 5,
      "coveredStatements": 6,
      "statementCov": 54.55,
      "statements": 11,
    },
    "orders/app/components/orders-subviews.js": Object {
      "conditionalCov": 94.74,
      "conditionals": 19,
      "coveredConditionals": 18,
      "coveredStatements": 17,
      "statementCov": 68,
      "statements": 25,
    },
  },
  "total": Object {
    "changed": Object {
      "conditionalCov": 80.19,
      "conditionals": 2489,
      "coveredConditionals": 1996,
      "coveredStatements": 3559,
      "statementCov": 87.19,
      "statements": 4082,
    },
    "diff": Object {
      "conditionalCov": 7.3,
      "conditionals": 10,
      "coveredConditionals": 190,
      "coveredStatements": 131,
      "statementCov": 2.3,
      "statements": 40,
    },
    "original": Object {
      "conditionalCov": 72.85,
      "conditionals": 2479,
      "coveredConditionals": 1806,
      "coveredStatements": 3428,
      "statementCov": 84.81,
      "statements": 4042,
    },
  },
}
`;
